sklearn/utils/_encode.py

```diff
@@ -27,13 +28,36 @@ def _unique(values, *, return_inverse=False):
     if values.dtype == object:
         return _unique_python(values, return_inverse=return_inverse)
     # numerical
-    return np.unique(values, return_inverse=return_inverse)
+    out = np.unique(values, return_inverse=return_inverse)
+
+    if return_inverse:
+        uniques, inverse = out
+    else:
+        uniques, inverse = out, None
+
+    # np.unique will have duplicate missing values at the end of `uniques`
+    # here we clip the nans and remove it from uniques
+    if uniques.size and is_scalar_nan(uniques[-1]):
+        nan_idx = np.searchsorted(uniques, np.nan)
+        uniques = uniques[:nan_idx + 1]
+        if return_inverse:
+            inverse[inverse > nan_idx] = nan_idx
+
+    if return_inverse:
+        return uniques, inverse
+    return uniques
 
 
 def _unique_python(values, *, return_inverse):
     # Only used in `_uniques`, see docstring there for details
     try:
-        uniques = sorted(set(values))
+        uniques_set = set(values)
+        missing_values_in_set = [value for value in (None, np.nan)
```

Comment:
is None not sortable?

Reply:
This won't match `float('nan')`, nor `(np.array(0)/0).item()`. Does it matter?

To be careful, I think you might be safer with:
```py
        missing_values_in_set = [value for value in uniques_set
                                 if value is None or is_scalar_nan(value)]
```